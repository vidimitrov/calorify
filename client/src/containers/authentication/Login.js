import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import logo from '../../assets/img/logo.png';
import Wrapper from '../../components/authentication/Wrapper';
import FormHeader from '../../components/authentication/FormHeader';
import FormContent from '../../components/authentication/FormContent';
import SecondaryAction from '../../components/authentication/SecondaryAction';
import Input from '../../components/authentication/Input';
import SubmitButton from '../../components/authentication/SubmitButton';
import Anchor from '../../components/authentication/Anchor';
import Image from '../../components/authentication/Image';
import ValidationLabel from '../../components/authentication/ValidationLabel';
import { login as loginActionCreator, resetErrors as resetErrorsAction } from '../../actions/authentication';
import { isValidEmail } from '../../lib/validations';

export class Login extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      email: '',
      validEmail: true,
      requiredEmail: false,
      password: '',
      requiredPassword: false,
    };

    this.onChangeHandler = this.onChangeHandler.bind(this);
  }

  onChangeHandler(key, value) {
    let { validEmail, requiredEmail, requiredPassword } = this.state;
    const { error, resetErrors } = this.props;

    if (error) {
      resetErrors();
    }

    if (key === 'email') {
      requiredEmail = false;
      if (!isValidEmail(value)) {
        validEmail = false;
      } else {
        validEmail = true;
      }
    }

    if (key === 'password') {
      requiredPassword = false;
    }

    this.setState({
      validEmail,
      requiredEmail,
      requiredPassword,
      [key]: value,
    });
  }

  render() {
    const { login, error, navigate } = this.props;
    const {
      email,
      password,
      validEmail,
      requiredEmail,
      requiredPassword,
    } = this.state;
    return (
      <Wrapper>
        <FormContent id="formContent">
          <FormHeader>
            <Image src={logo} alt="Calorify Logo" />
            <h1>Calorify</h1>
          </FormHeader>

          <div>
            <ValidationLabel
              invalid={error && error.statusCode === 400}
            >
              The email or password you entered is incorrect
            </ValidationLabel>
            <Input
              type="text"
              name="login"
              placeholder="Email"
              invalid={!validEmail || requiredEmail}
              onChange={e => this.onChangeHandler('email', e.target.value)}
            />
            <ValidationLabel invalid={!validEmail}>Email is incorrect</ValidationLabel>
            <ValidationLabel invalid={requiredEmail}>Email is required</ValidationLabel>
            <Input
              type="password"
              name="login"
              placeholder="Password"
              invalid={requiredPassword}
              onChange={e => this.onChangeHandler('password', e.target.value)}
            />
            <ValidationLabel invalid={requiredPassword}>Password is required</ValidationLabel>
            <SubmitButton
              disabled={!validEmail}
              onClick={() => {
                if (!email || !password) {
                  this.setState({
                    requiredEmail: !email,
                    requiredPassword: !password,
                  });
                } else {
                  login(email, password)
                    .then(() => {
                      navigate('/');
                    })
                    .catch(() => {
                      this.setState({
                        email: '',
                        password: '',
                        validEmail: true,
                      });
                    });
                }
              }}
            >
              Log In
            </SubmitButton>
            <SecondaryAction>
              <Anchor to="/auth/signup"> Create an account</Anchor>
            </SecondaryAction>
          </div>
        </FormContent>
      </Wrapper>
    );
  }
}

Login.propTypes = {
  login: PropTypes.func.isRequired,
  navigate: PropTypes.func.isRequired,
  resetErrors: PropTypes.func.isRequired,
  error: PropTypes.shape({
    message: PropTypes.string,
    statusCode: PropTypes.number,
    success: PropTypes.bool,
  }),
};

Login.defaultProps = {
  error: null,
};

const mapStateToProps = state => ({
  user: state.auth.user,
  error: state.auth.error,
});

const mapDispatchToProps = dispatch => ({
  login: (email, password) => dispatch(loginActionCreator(email, password)),
  resetErrors: () => dispatch(resetErrorsAction()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
