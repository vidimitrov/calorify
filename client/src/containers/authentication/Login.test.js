import React from 'react';
import renderer from 'react-test-renderer';
import { Login } from './Login';

it('Login container renders correctly', () => {
  const props = {
    login: () => { },
    resetErrors: () => { },
    navigate: () => { },
  };
  const tree = renderer.create(
    <Login {...props} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
