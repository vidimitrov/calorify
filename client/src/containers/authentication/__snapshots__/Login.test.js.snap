// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login container renders correctly 1`] = `
<div
  className="Wrapper-lhbnd-0 ipWKTt"
>
  <div
    className="FormContent-sc-7xu7iv-0 egRBsD"
    id="formContent"
  >
    <div
      className="FormHeader-la3hb0-0 gObcbi"
    >
      <img
        alt="Calorify Logo"
        className="Image-sc-1ufkk63-0 jvkVsO"
        src="logo.png"
      />
      <h1>
        Calorify
      </h1>
    </div>
    <div>
      <p
        className="ValidationLabel-ol62xd-0 jgOKXL"
      >
        The email or password you entered is incorrect
      </p>
      <input
        className="Input-sc-1i6e4g7-0 eABVIJ"
        name="login"
        onChange={[Function]}
        placeholder="Email"
        type="text"
      />
      <p
        className="ValidationLabel-ol62xd-0 jgOKXL"
      >
        Email is incorrect
      </p>
      <p
        className="ValidationLabel-ol62xd-0 jgOKXL"
      >
        Email is required
      </p>
      <input
        className="Input-sc-1i6e4g7-0 eABVIJ"
        name="login"
        onChange={[Function]}
        placeholder="Password"
        type="password"
      />
      <p
        className="ValidationLabel-ol62xd-0 jgOKXL"
      >
        Password is required
      </p>
      <button
        className="SubmitButton-x5zdfd-0 frSvnw"
        disabled={false}
        onClick={[Function]}
      >
        Log In
      </button>
      <div
        className="SecondaryAction-cirh9v-0 iBEjhO"
      >
        <a
          className="Anchor-sc-1c4e1zf-0 grrGrN"
          href="/auth/signup"
          onClick={[Function]}
        >
           Create an account
        </a>
      </div>
    </div>
  </div>
</div>
`;
